### ISSUE TITLE
Status line overflowing to window on its right

### ISSUE THREAD

**Issue by critiqjo** [neovim/neovim#4195]
Preview:
![overflow](https://cloud.githubusercontent.com/assets/1436441/12886007/7f0fff1c-ce90-11e5-9461-3b83fbe79283.png)

#### Minimal steps to reproduce
1. `nvim -u NONE`
2. Execute the following commands:

   ```vim
   :e /path/to/file1
   :20vsp /path/to/file2
   :let &statusline='%=%#Title#file: %{expand("%:p")} %#NonText# hello'
   ```

The right part of the status line needs to contain 2 (or more) different highlights. The second highlight onwards gets overflowed to the next window.

_Note:_ This is specific to Neovim; Vim has no such issues.

**Comment by critiqjo** [comment]
Just noticed that this could be a duplicate of #4147 but it is labeled `bug-vim`, which it is not. Also, a minimal way to reproduce it was not found in the discussion. Close either one, but the label should be just `bug`.

**Comment by justinmk** [comment]
The minimal steps are very useful, thanks.

**Comment by cHoco** [comment]
@critiqjo I think it's the same bug, I haven't investigated why it is more *visible* in neovim but could you check if this patch fixes it?
``` diff
diff --git a/src/nvim/buffer.c b/src/nvim/buffer.c
index a6e3fed..e0ebb30 100644
--- a/src/nvim/buffer.c
+++ b/src/nvim/buffer.c
@@ -3000,7 +3000,7 @@ int build_stl_str_hl(
           && item[groupitem[groupdepth]].minwid == 0) {
         bool has_normal_items = false;
         for (long n = groupitem[groupdepth] + 1; n < curitem; n++) {
-          if (item[n].type == Normal) {
+          if (item[n].type == Normal || item[n].type == Highlight) {
             has_normal_items = true;
             break;
           }
```
it's the same patch proposed by @chrisbra

**Comment by critiqjo** [comment]
Yep! It did! Thank you!

**Comment by gregorias** [comment]
@cHoco, I also suffer from this issue any your patch fixed it. Thanks!

**Comment by justinmk** [comment]
Can someone send a PR?

**Comment by cHoco** [comment]
Since this is a patch that has been submitted to Vim but hasn't been included yet, what's the usual process in this case?

Also since I still haven't read that function I don't feel qualified to submit this change :smile:

**Comment by svanharmelen** [comment]
Having the same issue, so having this patch in neovim asap would be really great!

**Comment by wsdjeg** [comment]
I have test this in vim,but can not reproduce, now only happened in neovim ,will be fixed?

**Comment by nicdumz** [comment]
OK, status of this:

- a vim upstream patch has been posted by @chrisbra ( https://groups.google.com/forum/#!msg/vim_dev/sb1jmVirXPU/mPhvDnZ-CwAJ ). Not yet merged.
- I've verified that the same patch (https://github.com/neovim/neovim/issues/4195#issuecomment-181920845) would fix the issue for neovim.

-> Should I send a similar patch for neovim, or do we want for upstream to catch up and then merge this patch in?

**Comment by wsdjeg** [comment]
https://github.com/neovim/neovim/issues/4195#issuecomment-181920845

this comment  fix this issue,after merge that comment ,but when I open the file ,I can not see status about syntastic,only after running the command `SyntasticCheck`,I can see it

**Comment by wsdjeg** [comment]
@justinmk it has been merged by vim_dev,
[vim-patch 7.4.1511](https://github.com/vim/vim/commit/af6e36ff16736106a1bc63bb4d01f51fdfeb29a2)

**Comment by ZyX-I** [comment]
@wsdjeg Why do you think you should see that status before syntastic actually checked anything? I guess you simply have automatic checking disabled.

**Comment by wsdjeg** [comment]
I do not want to see that status,I I have disable the check when I open a file.let me show you difference of what I see.
before this patch ,when I open a file ,the status line is
![2016-03-08 20-02-24](https://cloud.githubusercontent.com/assets/13142418/13601137/cb55917c-e568-11e5-84d0-d6e082be1eb2.png)
after this patch
![2016-03-08 20-05-10](https://cloud.githubusercontent.com/assets/13142418/13601184/1783ef9e-e569-11e5-9a9c-6b49fc81d22c.png)
there is a orange and red empty area,I think if there is nothing about syntastic,the area should be hidded

**Comment by justinmk** [comment]
>  Should I send a similar patch for neovim

PR welcome. Since the patch is in Vim please note https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim#pull-requests

**Comment by nicdumz** [comment]
Awesome, it indeed looks like the patch was merged a couple hours ago upstream.

See PR https://github.com/neovim/neovim/pull/4427

**Comment by apriendeau** [comment]
@nicdumz Nice work. This is not a big issue just an annoyance, it will be nice to have this fixed!

**Comment by nicdumz** [comment]
Merged, so I believe we can close this now.

**Comment by mhinz** [comment]
@critiqjo This should be fixed by https://github.com/neovim/neovim/pull/4427. Please reopen if that's not the case.

**Comment by wsdjeg** [comment]
@mhinz how about the picture I shown above,I think maybe it is an issue.

**Comment by mhinz** [comment]
@wsdjeg Can you reproduce with Vim and Neovim from master?

**Comment by wsdjeg** [comment]
both vim and neovim has this difference after that patch.maybe it is not the best selution

**Comment by mhinz** [comment]
@chrisbra Can you reproduce this? https://github.com/neovim/neovim/issues/4195#issuecomment-193756694

**Comment by cHoco** [comment]
It's not a issue with vim/neovim see here my comment for more info: https://github.com/vim-airline/vim-airline/issues/1037#issuecomment-193497502 
It's just a limitation of vim-airline, you can track progress here: https://github.com/vim-airline/vim-airline/issues/157

**Comment by mhinz** [comment]
:+1:

**Comment by wsdjeg** [comment]
:+1:
